### PR TITLE
Inject logger to WF-core

### PR DESCRIPTION
**Pull request by MatanYadaev**
## **Type**
enhancement, bug_fix

___

## **Description**
- Implements a logger interface and a function to set a custom logger in the workflow-core package.
- Migrates all instances of `console.log` and `console.error` to use the new logger interface across various plugins and utilities.
- Integrates the workflow-core logger with the application's logger service.
- Enforces a no console log rule in the ESLint configuration.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export Logger and SetLogger Functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/index.ts
- Exports `logger` and `setLogger` from the workflow-core package.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-148d287ac79cfaf247c628b3d8436f0564aac2d04dd0678374b33273c26536cd">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export Logger Utilities from Lib Index</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/index.ts
- Exports `logger` and `setLogger` from the logger module.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-e8b9e89ffc4d6e47e8e7293543075532b1c7d4aabeef5e895d0d1902d84d8031">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>logger.ts</strong><dd><code>Implement Logger Interface and SetLogger Function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/logger.ts
<li>Implements a default logger interface with console methods.<br> <li> Provides a <code>setLogger</code> function to override the default logger.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-79369d21d7ea96ea638c9e3df2a3e198007dec3854a4f162355ceeb3c43b0e7f">+21/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>app-logger.service.ts</strong><dd><code>Integrate Workflow-Core Logger with App Logger Service</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/common/app-logger/app-logger.service.ts
<li>Integrates workflow-core logger with the application logger service.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-55adf05f381da4a816b53b3a021eb41ea68ff389e9805e298cd4370a32449422">+9/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Bug_fix</strong></td><td><details><summary>7 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>iterative-plugin.ts</strong><dd><code>Migrate Iterative Plugin to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/plugins/common-plugin/iterative-plugin.ts
<li>Replaces <code>console.log</code> and <code>console.error</code> with the logger interface.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-cbc9b4dded358f941d7fa68389673220de7242441b9e0aa1b4b001be92c35bd7">+14/-11</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transformer-plugin.ts</strong><dd><code>Migrate Transformer Plugin to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/plugins/common-plugin/transformer-plugin.ts
- Replaces `console.log` with the logger interface.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-3a26a0c006651323740946ab47d531844cb8a35eb4a057d9724374a6430b079a">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>api-plugin.ts</strong><dd><code>Migrate API Plugin to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/plugins/external-plugin/api-plugin.ts
<li>Replaces <code>console.log</code> and <code>console.error</code> with the logger interface.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-ff1ae21f4192aaa50415159baf2f54727da9d5845efac07bfcb3cfbda7717dc6">+9/-4</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>email-plugin.ts</strong><dd><code>Migrate Email Plugin to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/plugins/external-plugin/email-plugin.ts
- Replaces `console.log` with the logger interface.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-cb30be7e5434424848a90966f681234ba8021e1137baae5da638e905964fac1d">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>webhook-plugin.ts</strong><dd><code>Migrate Webhook Plugin to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/plugins/external-plugin/webhook-plugin.ts
- Replaces `console.error` with the logger interface.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-aac0a0e450a3814ef4a76d3a9a104dba5034acc628cd09bee0e4e3e7b1751865">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>helpers-transformer.ts</strong><dd><code>Migrate Helpers Transformer to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/utils/context-transformers/helpers-transformer.ts
<li>Replaces <code>console.log</code> and <code>console.warn</code> with the logger interface.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-a962c53d6ea067eda25bce31100221fdfea1ebaae50a212cf34324e4130baee8">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>workflow-runner.ts</strong><dd><code>Migrate Workflow Runner to Use Logger</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/src/lib/workflow-runner.ts
<li>Replaces <code>console.log</code> and <code>console.error</code> with the logger interface.<br>

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-e885016468c29d2821f8c7ec44b4141a544cdb48b23622924cd2cccfcb7d3f46">+23/-11</a>&nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>.eslintrc.cjs</strong><dd><code>Enforce No Console Log Rule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/workflow-core/.eslintrc.cjs
- Enforces no console log rule in ESLint configuration.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-ad1627d77dd3691ace4228c99f6998a4131e7af74ce0628be7c0aa6066424dda">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>data-migrations</strong><dd><code>Update Data Migrations Submodule Reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/data-migrations
- Updates the data migrations submodule reference.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2222/files#diff-9c33eef7eba3ba7ca6006881973adb35d0cbc4048bb7256f945ab6da6f013edf">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: packages/workflow-core/src/lib/logger.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  debug(message: string, logData?: AnyRecord): void;
+};
+
+/* eslint-disable no-console */
+export let logger: LoggerInterface = {
+  log: (message: string, logData: AnyRecord = {}) => console.log(message, logData),
+  error: (message: string, logData: AnyRecord = {}) => console.error(message, logData),
+  warn: (message: string, logData: AnyRecord = {}) => console.warn(message, logData),
+  debug: (message: string, logData: AnyRecord = {}) => console.debug(message, logData),
+};
```
Comment (MatanYadaev) [inline packages/workflow-core/src/lib/logger.ts:9]: This is the default logger. Currently it's used in the `workflow-browser-sdk`

File: packages/workflow-core/src/lib/logger.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+/* eslint-enable no-console */
+
+export const setLogger = (newLogger: LoggerInterface) => {
+  logger = newLogger;
+};
```
Comment (MatanYadaev) [inline packages/workflow-core/src/lib/logger.ts:14]: I decided to make the logger a singleton in the WF-core, as it is already a singleton in the WF-service. It's easier than passing the logger when calling `createWorkflow` and then drilling it down to all other functions.

File: packages/workflow-core/src/lib/logger.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import { AnyRecord } from '@ballerine/common';
+
+export type LoggerInterface = {
+  log(message: string, logData?: AnyRecord): void;
+  error(message: string, logData?: AnyRecord): void;
+  warn(message: string, logData?: AnyRecord): void;
+  debug(message: string, logData?: AnyRecord): void;
+};
```
Comment (Omri-Levy) [inline packages/workflow-core/src/lib/logger.ts]: https://www.totaltypescript.com/method-shorthand-syntax-considered-harmful

**Review by Omri-Levy — APPROVED** [review]

File: packages/workflow-core/src/lib/logger.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  debug: (message: string, logData?: AnyRecord) => void;
+};
+
+/* eslint-disable no-console */
+export let logger: LoggerInterface = {
+  log: (message: string, logData: AnyRecord = {}) => console.log(message, logData),
+  error: (message: string, logData: AnyRecord = {}) => console.error(message, logData),
+  warn: (message: string, logData: AnyRecord = {}) => console.warn(message, logData),
+  debug: (message: string, logData: AnyRecord = {}) => console.debug(message, logData),
+};
```
Comment (liorzam) [inline packages/workflow-core/src/lib/logger.ts]: should be placed in common package

File: packages/workflow-core/src/lib/workflow-runner.ts [hunk 5]
```diff
@@ -212,7 +212,7 @@ export class WorkflowRunner {
     if (pluginKind === 'iterative') return IterativePlugin;
     if (pluginKind === 'transformer') return TransformerPlugin;
 
-    console.log('Plugin kind is not supplied or not supported, falling back to Iterative plugin.');
+    logger.log('Plugin kind is not supplied or not supported, falling back to Iterative plugin.');
```
Comment (liorzam) [inline packages/workflow-core/src/lib/workflow-runner.ts]:
```suggestion
    logger.log('Plugin kind is not supplied or not supported, falling back to Iterative plugin.', { pluginKind });
```

File: packages/workflow-core/src/lib/logger.ts (hunk 4 above)
Comment (MatanYadaev) [reply packages/workflow-core/src/lib/logger.ts]: Why?

Comment (MatanYadaev) [reply packages/workflow-core/src/lib/logger.ts]: The logger itself can't be in the common, as it's a singleton that is specific to `workflow-core`. The only thing that can be placed in `common` is `LoggerInterface`, and I'm not sure it's worth it.

**Comment by liorzam** [comment]
Please also run 'pnpm update-packages'

**Review by liorzam — APPROVED** [review]

**Comment by liorzam** [comment]
What do you think to add the functionality as i did on unified api that every call to 'error' will send it to sentry as well?

**Pull request merged**